flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/Dispatcher.java

```diff
@@ -350,66 +355,64 @@ private boolean isPartialResourceConfigured(JobGraph jobGraph) {
 
 	private CompletableFuture<Void> persistAndRunJob(JobGraph jobGraph) throws Exception {
 		jobGraphWriter.putJobGraph(jobGraph);
-
-		final CompletableFuture<Void> runJobFuture = runJob(jobGraph);
-
-		return runJobFuture.whenComplete(BiConsumerWithException.unchecked((Object ignored, Throwable throwable) -> {
-			if (throwable != null) {
-				jobGraphWriter.removeJobGraph(jobGraph.getJobID());
-			}
-		}));
+		return runJob(jobGraph, false);
 	}
 
-	private CompletableFuture<Void> runJob(JobGraph jobGraph) {
-		Preconditions.checkState(!jobManagerRunnerFutures.containsKey(jobGraph.getJobID()));
-
-		final CompletableFuture<JobManagerRunner> jobManagerRunnerFuture = createJobManagerRunner(jobGraph);
-
-		jobManagerRunnerFutures.put(jobGraph.getJobID(), jobManagerRunnerFuture);
-
-		return jobManagerRunnerFuture
-			.thenApply(FunctionUtils.uncheckedFunction(this::startJobManagerRunner))
-			.thenApply(FunctionUtils.nullFn())
-			.whenCompleteAsync(
-				(ignored, throwable) -> {
-					if (throwable != null) {
-						jobManagerRunnerFutures.remove(jobGraph.getJobID());
-					}
-				},
-				getMainThreadExecutor());
-	}
+	/**
+	 * @param blocking If true, the returned future will complete once the JobManager has been initialized
+	 */
+	private CompletableFuture<Void> runJob(JobGraph jobGraph, boolean blocking) {
+		Preconditions.checkState(!runningJobs.containsKey(jobGraph.getJobID()));
+		long jobManagerInitializationStarted = System.currentTimeMillis();
+		DispatcherJob dispatcherJob = new DispatcherJob();
+		runningJobs.put(jobGraph.getJobID(), dispatcherJob);
```

Comment:
Any reason this cannot be done after the JobManagerRunner and JobMasterGateway have been created?